native/dlaexc.go

```diff
@@ -0,0 +1,267 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dlaexc swaps two adjacent diagonal blocks of order 1 or 2 in an n×n upper
+// quasi-triangular matrix T by an orthogonal similarity transformation.
+//
+// T must be in Schur canonical form, that is, block upper triangular with 1×1
+// and 2×2 diagonal blocks; each 2×2 diagonal block has its diagonal elements
+// equal and its off-diagonal elements of opposite sign. On return, T will
+// contain the updated matrix again in Schur canonical form.
+//
+// If wantq is true, the transformation is accumulated in the n×n matrix Q,
+// otherwise Q is not referenced.
+//
+// j1 is the index of the first row of the first block. n1 and n2 are the order
+// of the first and second block, respectively.
+//
+// work much have length at least n, otherwise Dlaexc will panic.
+//
+// If ok is false, the transformed matrix T would be too far from Schur form.
+// The blocks are not swapped, and T and Q are not modified.
+func (impl Implementation) Dlaexc(wantq bool, n int, t []float64, ldt int, q []float64, ldq int, j1, n1, n2 int, work []float64) (ok bool) {
+	checkMatrix(n, n, t, ldt)
+	if wantq {
+		checkMatrix(n, n, q, ldq)
+	}
+	if j1 < 0 || n <= j1 {
+		panic("lapack: index j1 out of range")
+	}
+	if len(work) < n {
+		panic(badWork)
+	}
+	if n1 < 0 || 2 < n1 {
+		panic("lapack: invalid value of n1")
+	}
+	if n2 < 0 || 2 < n2 {
+		panic("lapack: invalid value of n2")
+	}
+
+	if n == 0 || n1 == 0 || n2 == 0 {
+		return true
+	}
+	if j1+n1 >= n {
+		return true
+	}
+
+	j2 := j1 + 1
+	j3 := j1 + 2
+
+	bi := blas64.Implementation()
+
+	if n1 == 1 && n2 == 1 {
+		// Swap two 1×1 blocks.
+
+		t11 := t[j1*ldt+j1]
+		t22 := t[j2*ldt+j2]
+
+		// Determine the transformation to perform the interchange.
+		cs, sn, _ := impl.Dlartg(t[j1*ldt+j2], t22-t11)
+
+		// Apply transformation to the matrix T.
+		if j3 < n {
+			bi.Drot(n-j3, t[j1*ldt+j3:], 1, t[j2*ldt+j3:], 1, cs, sn)
+		}
+		if j1 > 0 {
+			bi.Drot(j1, t[j1:], ldt, t[j2:], ldt, cs, sn)
+		}
+
+		t[j1*ldt+j1] = t22
+		t[j2*ldt+j2] = t11
+
+		if wantq {
+			// Accumulate transformation in the matrix Q.
+			bi.Drot(n, q[j1:], ldq, q[j2:], ldq, cs, sn)
+		}
+
+		return true
+	}
+
+	// Swapping involves at least one 2×2 block.
+	//
+	// Copy the diagonal block of order n1+n2 to the local array d and
+	// compute its norm.
+	nd := n1 + n2
+	ldd := nd
+	d := make([]float64, nd*nd)
```

Comment:
If by asking you mean something like `go build -gcflags -S 2>&1 | less`, then I asked. When `d` and `x` are created with `make`, there are 2 calls to `runtime.makeslice` in the output. When they are local arrays, `[16]float64` and `[4]float64`, respectively, there is one call to `runtime.newobject` for `d` but no such call for `x`. For `x` there is instead
```
MOVQ    $0, "".x+432(SP)
MOVQ    $0, "".x+440(SP)
MOVQ    $0, "".x+448(SP)
MOVQ    $0, "".x+456(SP)
```
`x` is obviously on the stack, I don't know about `d`, maybe allocated? Interestingly, there are also calls to `newobject` for each of `u`, `u1`, and `u2` below that are shorter than `x`.

I've changed both `d` and `x` to arrays.